File: name-validation-2-nodejs/REST/name_info_v2_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { NameInfoV2Response } from './nv2_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Name Validation 2 API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/NV2/api.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Name Validation 2 API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/NV2/api.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Name Validation 2 API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/NV2/api.svc/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * <summary>
 * Constructs a full URL for the NameInfoV2 API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}NameInfoV2?${querystring.stringify(params)}&format=json`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<NameInfoV2Response>">A promise that resolves to a NameInfoV2Response object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new NameInfoV2Response(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Name Validation 2 API's NameInfoV2 endpoint,
 * retrieving name validation information with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const NameInfoV2Client = {
    /**
     * <summary>
     * Asynchronously invokes the NameInfoV2 API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * </summary>
     * @param {string} Name - The name to validate.
     * @param {string} Option - Comma-separated list of options for additional processing (optional).
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<NameInfoV2Response>} - A promise that resolves to a NameInfoV2Response object.
     */
    async invokeAsync(Name, Option = '', LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Name,
            Option,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return isValid(fallbackResponse) ? fallbackResponse : response;
        }

        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the NameInfoV2 API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @returns {NameInfoV2Response} - A NameInfoV2Response object with name validation details or an error.
     */
    invoke(Name, Option = '', LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Name, Option, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { NameInfoV2Client, NameInfoV2Response };

File: name-validation-2-python/REST/name_info_v2_rest.py

import requests
from nv2_response import NameInfoV2Response, NameInfoV2, BestGuessName, Error

# Endpoint URLs for Name Validation 2 NameInfoV2 REST API
primary_url = 'https://sws.serviceobjects.com/NV2/api.svc/NameInfoV2?'
backup_url = 'https://swsbackup.serviceobjects.com/NV2/api.svc/NameInfoV2?'
trial_url = 'https://trial.serviceobjects.com/NV2/api.svc/NameInfoV2?'

def get_name_info_v2(
    name: str,
    option: str,
    license_key: str,
    is_live: bool = True,
    timeout_seconds: int = 15
) -> NameInfoV2Response:
    """
    Call Name Validation 2 NameInfoV2 API to retrieve name validation information.

    Parameters:
        name: The name to validate.
        option: Comma-separated list of options for additional processing (optional).
        license_key: Your license key to use the service.
        is_live: Value to determine whether to use the live or trial servers (default: True).
        timeout_seconds: Timeout, in seconds, for the call to the service (default: 15).

    Returns:
        NameInfoV2Response: Parsed JSON response with name information or error details.
    """
    params = {
        'Name': name,
        'Option': option,
        'LicenseKey': license_key,
        'format': 'json' 
    }

    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    # Attempt primary (or trial) endpoint first
    try:
        response = requests.get(url, params=params, timeout=timeout_seconds)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = getattr(response, 'Error', None)
        if not (error is None or getattr(error, 'TypeCode', None) != "3"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                data = response.json()

            # If still error, propagate exception
            if 'Error' in data:
                raise RuntimeError(f"NV2 service error: {data['Error']}")

            else:
                # Trial mode error is terminal
                raise RuntimeError(f"NV2 trial error: {data['Error']}")

        # Convert JSON response to NameInfoV2Response for structured access
        error = Error(**data.get('Error', {})) if data.get('Error') else None
        name_info_v2 = None
        if data.get('NameInfoV2'):
            best_guess_name = BestGuessName(**data['NameInfoV2'].get('BestGuessName', {})) if data['NameInfoV2'].get('BestGuessName') else None
            name_info_v2 = NameInfoV2(
                BestGuessName=best_guess_name,
                NameIn=data['NameInfoV2'].get('NameIn'),
                NameClassification=data['NameInfoV2'].get('NameClassification'),
                Prefix=data['NameInfoV2'].get('Prefix'),
                FirstName=data['NameInfoV2'].get('FirstName'),
                MiddleName=data['NameInfoV2'].get('MiddleName'),
                LastName=data['NameInfoV2'].get('LastName'),
                Suffix=data['NameInfoV2'].get('Suffix'),
                FirstNameFound=data['NameInfoV2'].get('FirstNameFound'),
                IsCommonFirstName=data['NameInfoV2'].get('IsCommonFirstName'),
                FirstNameOrigin=data['NameInfoV2'].get('FirstNameOrigin'),
                FirstNameSimilar=data['NameInfoV2'].get('FirstNameSimilar'),
                LastNameFound=data['NameInfoV2'].get('LastNameFound'),
                IsCommonLastName=data['NameInfoV2'].get('IsCommonLastName'),
                LastNameOrigin=data['NameInfoV2'].get('LastNameOrigin'),
                LastNameSimilar=data['NameInfoV2'].get('LastNameSimilar'),
                Gender=data['NameInfoV2'].get('Gender'),
                FirstNameAlt=data['NameInfoV2'].get('FirstNameAlt'),
                MiddleNameAlt=data['NameInfoV2'].get('MiddleNameAlt'),
                LastNameAlt=data['NameInfoV2'].get('LastNameAlt'),
                FirstNameAltFound=data['NameInfoV2'].get('FirstNameAltFound'),
                LastNameAltFound=data['NameInfoV2'].get('LastNameAltFound'),
                GenderAlt=data['NameInfoV2'].get('GenderAlt'),
                RelatedNames=data['NameInfoV2'].get('RelatedNames'),
                IsCorrectedName=data['NameInfoV2'].get('IsCorrectedName'),
                IsBusinessName=data['NameInfoV2'].get('IsBusinessName'),
                BusinessName=data['NameInfoV2'].get('BusinessName'),
                VulgarityScore=data['NameInfoV2'].get('VulgarityScore'),
                CelebrityScore=data['NameInfoV2'].get('CelebrityScore'),
                BogusScore=data['NameInfoV2'].get('BogusScore'),
                GarbageScore=data['NameInfoV2'].get('GarbageScore'),
                FirstNameDictionaryScore=data['NameInfoV2'].get('FirstNameDictionaryScore'),
                MiddleNameDictionaryScore=data['NameInfoV2'].get('MiddleNameDictionaryScore'),
                LastNameDictionaryScore=data['NameInfoV2'].get('LastNameDictionaryScore'),
                OverallNameScore=data['NameInfoV2'].get('OverallNameScore'),
                IsNameGood=data['NameInfoV2'].get('IsNameGood'),
                StatusCodes=data['NameInfoV2'].get('StatusCodes'),
                Status=data['NameInfoV2'].get('Status')
            )

        return NameInfoV2Response(
            NameInfoV2=name_info_v2,
            Error=error
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                data = response.json()
                if 'Error' in data:
                    raise RuntimeError(f"NV2 backup error: {data['Error']}") from req_exc

                # Convert JSON response to NameInfoV2Response for structured access
                error = Error(**data.get('Error', {})) if data.get('Error') else None
                name_info_v2 = None
                if data.get('NameInfoV2'):
                    best_guess_name = BestGuessName(**data['NameInfoV2'].get('BestGuessName', {})) if data['NameInfoV2'].get('BestGuessName') else None
                    name_info_v2 = NameInfoV2(
                        BestGuessName=best_guess_name,
                        NameIn=data['NameInfoV2'].get('NameIn'),
                        NameClassification=data['NameInfoV2'].get('NameClassification'),
                        Prefix=data['NameInfoV2'].get('Prefix'),
                        FirstName=data['NameInfoV2'].get('FirstName'),
                        MiddleName=data['NameInfoV2'].get('MiddleName'),
                        LastName=data['NameInfoV2'].get('LastName'),
                        Suffix=data['NameInfoV2'].get('Suffix'),
                        FirstNameFound=data['NameInfoV2'].get('FirstNameFound'),
                        IsCommonFirstName=data['NameInfoV2'].get('IsCommonFirstName'),
                        FirstNameOrigin=data['NameInfoV2'].get('FirstNameOrigin'),
                        FirstNameSimilar=data['NameInfoV2'].get('FirstNameSimilar'),
                        LastNameFound=data['NameInfoV2'].get('LastNameFound'),
                        IsCommonLastName=data['NameInfoV2'].get('IsCommonLastName'),
                        LastNameOrigin=data['NameInfoV2'].get('LastNameOrigin'),
                        LastNameSimilar=data['NameInfoV2'].get('LastNameSimilar'),
                        Gender=data['NameInfoV2'].get('Gender'),
                        FirstNameAlt=data['NameInfoV2'].get('FirstNameAlt'),
                        MiddleNameAlt=data['NameInfoV2'].get('MiddleNameAlt'),
                        LastNameAlt=data['NameInfoV2'].get('LastNameAlt'),
                        FirstNameAltFound=data['NameInfoV2'].get('FirstNameAltFound'),
                        LastNameAltFound=data['NameInfoV2'].get('LastNameAltFound'),
                        GenderAlt=data['NameInfoV2'].get('GenderAlt'),
                        RelatedNames=data['NameInfoV2'].get('RelatedNames'),
                        IsCorrectedName=data['NameInfoV2'].get('IsCorrectedName'),
                        IsBusinessName=data['NameInfoV2'].get('IsBusinessName'),
                        BusinessName=data['NameInfoV2'].get('BusinessName'),
                        VulgarityScore=data['NameInfoV2'].get('VulgarityScore'),
                        CelebrityScore=data['NameInfoV2'].get('CelebrityScore'),
                        BogusScore=data['NameInfoV2'].get('BogusScore'),
                        GarbageScore=data['NameInfoV2'].get('GarbageScore'),
                        FirstNameDictionaryScore=data['NameInfoV2'].get('FirstNameDictionaryScore'),
                        MiddleNameDictionaryScore=data['NameInfoV2'].get('MiddleNameDictionaryScore'),
                        LastNameDictionaryScore=data['NameInfoV2'].get('LastNameDictionaryScore'),
                        OverallNameScore=data['NameInfoV2'].get('OverallNameScore'),
                        IsNameGood=data['NameInfoV2'].get('IsNameGood'),
                        StatusCodes=data['NameInfoV2'].get('StatusCodes'),
                        Status=data['NameInfoV2'].get('Status')
                    )

                return NameInfoV2Response(
                    NameInfoV2=name_info_v2,
                    Error=error
                )
            except Exception as backup_exc:
                raise RuntimeError("NV2 service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"NV2 trial error: {str(req_exc)}") from req_exc